### PR TITLE
Issue #171 Improve application lifecycle management

### PR DESCRIPTION
**Pull request by az-qbradley**

### CODE REVIEW

File: squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala [hunk 1]
```diff
@@ -38,4 +38,14 @@ object Shutdown extends App {
   val preConfig = UnicomplexBoot.getFullConfig(None)
   val actorSystemName = preConfig.getString("squbs.actorsystem-name")
   Unicomplex(actorSystemName) ! GracefulStop
+}
+
+class JvmShutdownHook extends ExtensionLifecycle {
+  override def init(): Unit = {
+    Runtime.getRuntime.addShutdownHook(new Thread() {
```
Comment (akara) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala]: Codacy Issue found: [Avoid creating Threads](https://www.codacy.com/app/akara-sucharitakul/squbs/file/1873434557/issues/source?bid=2834325&fileBranchId=3087814#l45)

Comment (akara) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala]: You can avoid this Codacy issue by using the Scala version of shutdown hook. Please see http://www.scala-lang.org/api/2.11.7/index.html#scala.sys.package

File: docs/packaging.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+- Copy either of the archives to the running instance
+- `<app>-<version>.tar.gz`
+- `<app>-<version>.zip`
+- For example, explode the tarball `tar zxvf <app>-<version>.tar.gz` to the `<app>-<version>` directory
+- start the application `<app>-<version>/bin/run &`
+- You can check the admin `http://localhost:8080/adm` from that instance, or `http://<host>:8080/adm`
+
+## Shutdown
+
+You can terminate the running process, for example, in linux `kill $(lsof -ti TCP:8080 | head -1)`
```
Comment (akara) [inline docs/packaging.md]: Couple of thoughts here:
1. I assume you have tested this with very long shutdown times (minutes) and it still works well.
2. We need to document the signals in this case. The default signal in this example is SIGTERM. I don't expect it to wait with SIGKILL. Now, what if other signals get sent to the JVM?

Comment (az-qbradley) [inline docs/packaging.md]: I try not to imply any specific shutdown process here. User would assume JVM signal handling themselves for all env (linux, mac, windows ...). All we do here, is to recognize JVM signal to pass on-to Unicomplex.
Yes, I tested it. But, if things start to hang, they would need another "kill -9" or something to the process.

File: squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala (hunk 1 above)
Comment (az-qbradley) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala]: let me check

Comment (az-qbradley) [inline squbs-unicomplex/src/main/scala/org/squbs/unicomplex/Bootstrap.scala]: Cool, fixed.

File: docs/packaging.md (hunk 2 above)
Comment (az-qbradley) [inline docs/packaging.md]: Let me try something ...
See if I can block the shutdown thread (Await) for the Lifecycle "Stopped" or up-to a configurable timeout

Comment (akara) [inline docs/packaging.md]: @az-qbradley Did you get to try the latter? If you're all good I'm good to merge.

Comment (az-qbradley) [inline docs/packaging.md]: Yes, I added that part in ... https://github.com/paypal/squbs/pull/174/files#diff-bc5681ab3fccc9d4c3db904cf6bad706R56
Line #56 - 58

**Pull request merged**